File: examples/python/send_fcdp.py
#!/usr/bin/env python3
"""Send a valid FCDP v0.1 test access unit without any third-party package."""
import socket
import struct
import sys

HEADER_LEN = 37

def crc16(data: bytes) -> int:
    value = 0xFFFF
    for byte in data:
        value ^= byte << 8
        for _ in range(8):
            value = ((value << 1) ^ 0x1021) & 0xFFFF if value & 0x8000 else (value << 1) & 0xFFFF
    return value

def packet(payload: bytes, sequence: int = 1, frame: int = 1) -> bytes:
    if len(payload) > 1163:
        raise ValueError("sample supports one FCDP fragment only")
    header = b"FC" + bytes([1, 3, 0, 0])
    header += struct.pack(">QHHIIHHBHH", 1, 1, 1, sequence, frame, 0, 1, 0, 1000, len(payload))
    return header + struct.pack(">H", crc16(header)) + payload

if __name__ == "__main__":
    host, port, text = sys.argv[1], int(sys.argv[2]), sys.argv[3]
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.sendto(packet(text.encode()), (host, port))

File: examples/node/send-fcdp.mjs
// Send a valid FCDP v0.1 test access unit using only Node's standard library.
import dgram from "node:dgram";

function crc16(buffer) {
  let value = 0xffff;
  for (const byte of buffer) {
    value ^= byte << 8;
    for (let bit = 0; bit < 8; bit += 1) value = value & 0x8000 ? ((value << 1) ^ 0x1021) & 0xffff : (value << 1) & 0xffff;
  }
  return value;
}

function packet(text) {
  const payload = Buffer.from(text, "utf8");
  if (payload.length > 1163) throw new Error("sample supports one FCDP fragment only");
  const header = Buffer.alloc(35);
  header.write("FC", 0, "ascii"); header[2] = 1; header[3] = 3;
  header.writeBigUInt64BE(1n, 6); header.writeUInt16BE(1, 14); header.writeUInt16BE(1, 16);
  header.writeUInt32BE(1, 18); header.writeUInt32BE(1, 22); header.writeUInt16BE(0, 26);
  header.writeUInt16BE(1, 28); header[30] = 0; header.writeUInt16BE(1000, 31); header.writeUInt16BE(payload.length, 33);
  const checksum = Buffer.alloc(2); checksum.writeUInt16BE(crc16(header));
  return Buffer.concat([header, checksum, payload]);
}

const [host, port, text] = process.argv.slice(2);
if (!host || !port || text === undefined) throw new Error("usage: node send-fcdp.mjs <host> <port> <text>");
const socket = dgram.createSocket("udp4");
socket.send(packet(text), Number(port), host, () => socket.close());
